test/e2e/networking_perf.go

```diff
@@ -0,0 +1,238 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+// Tests network performance using iperf or other containers.
+import (
+	"fmt"
+	"math"
+	"strconv"
+	"strings"
+	"time"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/pkg/util/intstr"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+// Declared as Flakey since it has not been proven to run in parallel on small nodes or slow networks in CI
+var _ = Describe("Networking IPerf [Experimental] [Networking] [Flaky]", func() {
+	gceBandwidthBits := int64(30000000000)
+
+	f := NewDefaultFramework("network-perf")
+	run(f, 1, 1, gceBandwidthBits)
+	run(f, 1, 5, gceBandwidthBits)
+	run(f, 1, 8, gceBandwidthBits)
+	// as many clients as possible based on node count.
+	run(f, 1, 10000, gceBandwidthBits)
+})
+
+func run(f *Framework, numServer int, numClient int, maxBandwidthBits int64) {
+	It(fmt.Sprintf("should transfer ~ 1GB onto the service endpoint %v servers (maximum of %v clients)", numServer, numClient), func() {
+		nodes := ListSchedulableNodesOrDie(f.Client)
+		totalPods := len(nodes.Items)
+		// for a single service, we expect to divide bandwidth between the network.  Very crude estimate.
+		expectedBandwidth := int(float64(maxBandwidthBits) / float64(totalPods))
+		Expect(totalPods).NotTo(Equal(0))
+		appName := "iperf-e2e"
+		CreateServiceWithPods(f,
+			8001,
+			8002,
+			appName,
+			api.Container{
+				Name:  "iperf-server",
+				Image: "mattgruter/iperf",
+				Args: []string{
+					"-s", "-p", "8001"},
+				Ports: []api.ContainerPort{{ContainerPort: 8001}},
+			},
+			// this will be used to generate the -service name which all iperf clients point at.
+			numServer, // 1 server
+		)
+		CreateServiceWithPods(f,
```

Comment:
we're manually writing a pod to each node, for up to N nodes max, so its "sort of" like a replication controller but not really.